test/Main.hs

```diff
@@ -73,6 +82,58 @@ spec = do
         dpiTimeStampToUTCDPITimeStamp dpi `shouldBe` expected
 
       it "Should roundtrip UTCTime through DPITimestamp (w/ nanos -- not picos) " $ \_ -> do
-        property $ \tod day (nanos :: Nano) -> do
-          let utc = UTCTime day $ timeOfDayToTime tod { todSec = realToFrac nanos }
+        property $ \(UTCTimeNanos utc) -> do
           utc `shouldBe` dpiTimeStampToUTCTime (utcTimeToDPITimestamp utc)
+
+    describe "Roundtrip tests" $ do
+      it "Should round trip random values from a table" $ \conn -> do
+        property $ \x@TestTable{..} -> do
+          execute_ conn "create table test (a varchar(300), b number (12,0), c number (12,0), d number (12,0) null, e timestamp, f number (38,28))"
+          execute conn "insert into test values (:1,:2,:3,:4,:5,:6)" x
+          y  <- query_ conn "select * from test"
+          execute_ conn "drop table test"
+          [x] `shouldBe` y
+
+data TestTable
+  = TestTable
+  { fieldText :: Text
+  , fieldInt :: Int
+  , fieldInt64 :: Int64
+  , fieldMaybeInt :: Maybe Int
+  , fieldUTCTime :: UTCTimeNanos
+  , fieldDouble :: Double
+  } deriving stock (Generic, Show)
+    deriving anyclass (ToRow, FromRow)
+
+instance Eq TestTable where
+  x == y =
+    and
+    [ ((==) `on` fieldText) x y
+    , ((==) `on` fieldInt) x y
+    , ((==) `on` fieldInt64) x y
+    , ((==) `on` fieldMaybeInt) x y
+    , ((==) `on` fieldUTCTime) x y
+    , ((~==) `on` fieldDouble) x y
+    ]
+
+instance Arbitrary TestTable where
+  arbitrary =
+    TestTable
+      <$> arbitrary
+      <*> arbitrary -- choose (- 2^12, 2 ^ 12)
```

Comment:
The `number(12,0)` fields above have a range of `[999999999999, -999999999999]`, these tests fail with a "value larger than specified precision" error because Int/Int64 are wider.